File: G/sol.py
MAXN = 52
W = [97, 1007]
MOD = [10**9 + 7, 10**9 + 33]

prefix_hash = [[], []] # [2][2][MAXN][MAXN];
WPOW = [] # [2][MAXN][MAXN];

class hash_t:
    def __init__(self):
        self.har = [0, 0]
        self.height = 0
        self.width = 0

    def __eq__(self, rhs):
        if self.har != rhs.har:
            return False

        return self.height == rhs.height  and  self.width == rhs.width

    def __hash__(self):
        return hash(tuple(self.har) + (self.height, self.width))


def getmod(x, mod):
   tmp = x % mod
   return tmp + mod if x < 0 else tmp


def make_wpow():
    for hh in range(len(W)):
        w = W[hh]
        mod = MOD[hh]
        cur = 1
        wp0 = []
        for i in range(MAXN):
            wp1 = []
            for j in range(MAXN):
                wp1.append(cur)
                cur = cur * w % mod
            wp0.append(wp1)
        WPOW.append(wp0)


def make_prefix_hashes(s, height, width, dest_h):
    for hh in range(len(W)):
        dest_h.append([])
        colhash = [0 for i in range(width)]
        mod = MOD[hh]
        for i in range(height):
            dest_h[hh].append([])
            for j in range(width):
                colhash[j] = (colhash[j] + ord(s[i][j]) * WPOW[hh][i][j] % mod)\
                        % mod
                dest_h[hh][i].append((
                        (dest_h[hh][i][-1] if j else 0) + colhash[j]) % mod)


def get_subhash(h, luy, lux, rdy, rdx):
    res = hash_t()
    for hh in range(len(W)):
        mod = MOD[hh]

        pref_all = h[hh][rdy][rdx];
        pref_up = h[hh][luy - 1][rdx] if luy > 0 else 0
        pref_left = h[hh][rdy][lux - 1] if lux > 0 else 0
        pref_up_left = h[hh][luy - 1][lux - 1] if luy > 0 and lux > 0 else 0

        subsq = getmod(getmod(getmod(
                pref_all - pref_up , mod) - pref_left, mod) + pref_up_left, mod)
        moved = subsq * WPOW[hh][MAXN - rdy - 1][MAXN - rdx - 1] % mod

        res.har[hh] = moved

    res.height = rdy - luy + 1
    res.width = rdx - lux + 1
    return res;


def check(height, width, sq_height, sq_width, res):
    used = {}
    for luy in range(height[0] - sq_height, -1, -1):
        for lux in range(width[0] - sq_width, -1, -1):
            rdy = luy + sq_height - 1
            rdx = lux + sq_width - 1
            curhash = get_subhash(prefix_hash[0], luy, lux, rdy, rdx)
            used[curhash] = (luy, lux)

    for luy in range(height[1] - sq_height, -1, -1):
        for lux in range(width[1] - sq_width, -1, -1):
            rdy = luy + sq_height - 1
            rdx = lux + sq_width - 1
            curhash = get_subhash(prefix_hash[1], luy, lux, rdy, rdx)
            if curhash in used:
                res[0] = used[curhash]
                res[1] = (luy, lux)
                return True

    return False


def main():
    s = []
    height, width = [0, 0], [0, 0]

    for iter in range(2):
        height[iter], width[iter] = [int(x) for x in input().strip().split(' ')]
        s.append([])
        for i in range(height[iter]):
            s[iter].append(input().strip())

    make_wpow()

    for i in range(2):
        make_prefix_hashes(s[i], height[i], width[i], prefix_hash[i])

    min_h, min_w = min(height), min(width)

    ans_sq = 0;
    anspt = [None, None]
    ans_hw = (None, None)

    for h in range(min_h, 0, -1):
        bspt = [None, None]
        l, r = 1, min_w
        while l + 1 < r:
            m = (l + r) // 2
            if h * m > ans_sq and check(height, width, h, m, bspt):
                l = m
            else:
                r = m - 1

        if l < r and h * r > ans_sq and check(height, width, h, r, bspt):
            l = r
        if h * l > ans_sq and check(height, width, h, l, bspt):
            ans_sq = h * l
            anspt = bspt
            ans_hw = (h, l)

    if ans_sq:
        print(ans_hw[0], ans_hw[1])
        for pt in anspt:
            print(pt[0] + 1, pt[1] + 1)
    else:
        print(0, 0)

main()

File: G/sol.js
const readline = require('readline');
const rl = readline.createInterface({
    input: process.stdin,
    terminal: false
});

var MAXN = 52;
var W = [97, 1007];
MOD = [40484569, 57985519];

var prefix_hash = [[], []]; // [2][2][MAXN][MAXN];
var WPOW = []; // [2][MAXN][MAXN];

function hash_t() {
    this.har = [0, 0];
    this.height = 0;
    this.width = 0;
}

hash_t.prototype.str = function() {
    return this.har[0] + ';' + this.har[1] + ';' +
        this.height + ';' + this.width;
}


function getmod(x, mod) {
    tmp = x % mod;
    return x < 0 ? tmp + mod : tmp;
}


function make_wpow() {
    for (var hh = 0; hh < W.length; ++hh) {
        var w = W[hh];
        var mod = MOD[hh];
        var cur = 1;
        var wp0 = [];
        for (var i = 0; i < MAXN; ++i) {
            var wp1 = [];
            for (var j = 0; j < MAXN; ++j) {
                wp1.push(cur);
                cur = cur * w % mod;
            }
            wp0.push(wp1);
        }
        WPOW.push(wp0);
    }
}


function make_prefix_hashes(s, height, width, dest_h) {
    for (var hh = 0; hh < W.length; ++hh) {
        dest_h.push([]);
        var colhash = new Array(width);
        for (var i = 0; i < width; ++i)
            colhash[i] = 0;
        var mod = MOD[hh];
        for (var i = 0; i < height; ++i) {
            dest_h[hh].push([]);
            for (var j = 0; j < width; ++j) {
                colhash[j] = (colhash[j] + s[i].charCodeAt(j) * WPOW[hh][i][j] % mod) % mod;
                dest_h[hh][i].push(((j ? dest_h[hh][i][j - 1] : 0) + colhash[j]) % mod);
            }
        }
    }
}


function get_subhash(h, luy, lux, rdy, rdx) {
    var res = new hash_t();
    for (var hh = 0; hh < W.length; ++hh) {
        var mod = MOD[hh];

        var pref_all = h[hh][rdy][rdx];
        var pref_up = luy > 0 ? h[hh][luy - 1][rdx] : 0;
        var pref_left = lux > 0 ? h[hh][rdy][lux - 1] : 0;
        var pref_up_left = luy > 0 && lux > 0 ? h[hh][luy - 1][lux - 1] : 0;

        var subsq = getmod(getmod(getmod(
                pref_all - pref_up , mod) - pref_left, mod) + pref_up_left, mod);
        var moved = subsq * WPOW[hh][MAXN - rdy - 1][MAXN - rdx - 1] % mod;

        res.har[hh] = moved;
    }

    res.height = rdy - luy + 1;
    res.width = rdx - lux + 1;
    return res;
}


function check(height, width, sq_height, sq_width, res) {
    var used = {};
    for (var luy = height[0] - sq_height; luy >= 0; --luy)
        for (var lux = width[0] - sq_width; lux >= 0; --lux) {
            var rdy = luy + sq_height - 1;
            var rdx = lux + sq_width - 1;
            var curhash = get_subhash(prefix_hash[0], luy, lux, rdy, rdx);
            used[curhash.str()] = [luy, lux];
        }

    for (var luy = height[1] - sq_height; luy >= 0; --luy)
        for (var lux = width[1] - sq_width; lux >= 0; --lux) {
            var rdy = luy + sq_height - 1;
            var rdx = lux + sq_width - 1;
            var curhash = get_subhash(prefix_hash[1], luy, lux, rdy, rdx);
            var chs = curhash.str();
            if (chs in used) {
                res[0] = used[chs];
                res[1] = [luy, lux];
                return true;
            }
        }

    return false;
}


var s = [];
var height = [], width = [];


function solve() {
    make_wpow();

    for (var i = 0; i < 2; ++i)
        make_prefix_hashes(s[i], height[i], width[i], prefix_hash[i]);

    var min_h = Math.min(height[0], height[1]);
    var min_w = Math.min(width[0], width[1]);

    var ans_sq = 0;
    var anspt = [null, null];
    var ans_hw = [null, null];

    for (var h = min_h; h >= 1; --h) {
        var bspt = [null, null];
        var l = 1, r = min_w;
        while (l + 1 < r) {
            var m = (l + r) >> 1;
            if (h * m > ans_sq && check(height, width, h, m, bspt))
                l = m;
            else
                r = m - 1;
        }

        if (l < r && h * r > ans_sq && check(height, width, h, r, bspt))
            l = r;
        if (h * l > ans_sq && check(height, width, h, l, bspt)) {
            var ans_sq = h * l;
            var anspt = bspt;
            var ans_hw = [h, l];
        }
    }

    if (ans_sq) {
        console.log(ans_hw[0], ans_hw[1]);
        console.log(anspt[0][0] + 1, anspt[0][1] + 1);
        console.log(anspt[1][0] + 1, anspt[1][1] + 1);
    }
    else
        console.log(0, 0);
}

var iter = 0, ln;
rl.on('line', function (data) {
    if (iter < 2) {
        if (height.length == iter) {
            var x = data.trim().split(' ')
            height.push(parseInt(x[0]));
            width.push(parseInt(x[1]));
            s.push([]);
            ln = 0;
            return;
        }
        s[iter].push(data.trim());
        ++ln;
        if (ln == height[iter])
            ++iter;
        if (iter == 2)
            solve();
    }
});

